Home Screen - migrated orders panel

octaedro opened the pull request:
Partially fixes [5238](https://github.com/woocommerce/woocommerce-admin/issues/5238)

This PR moves the Orders panel to the `Home` screen. In order to do it, an ActivityPanel component is created in the `Home` screen.
A follow-up PR will be created with the changes and additions necessary to finish up the issue. Those changes are the `Manage all orders` button and the empty activity panel for orders.


### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots
![Screen Capture on 2020-10-22 at 10-43-10](https://user-images.githubusercontent.com/1314156/96880354-756f4680-1453-11eb-99c5-4968a755151f.gif)


### Detailed test instructions:

- Go to the `Home` screen.
- Verify the orders panel is visible.
- Create a few orders (manually or using the [Smooth generator](https://github.com/woocommerce/wc-smooth-generator)).
- Confirm that the orders in progress number is right.
- Verify that the order card looks like it should.
- Verify the buttons and links are working well.

<!--- Note: When displaying information based on sample data, such as SwaggerHub, 
be sure to detail parts affected in Release Notes --->

### Changelog Note:
Dev: Home Screen - migrate orders panel
<!--- Optional: Enter a changelog note following the WooCommerce core format using prefixes of Enhancement:, Tweak:, Dev:, Fix:, Performance:. If no note is entered, one will be constructed from the title and labels. --->

octaedro committed:
Removed "getUnreadOrders" from "unread-indicators.js"

This commit removes the method "getUnreadOrders" from "unread-indicators.js"

octaedro committed:
Orders panel moved

octaedro committed:
Created "ActivityPanel" component

octaedro committed:
Added activity panel to home screen

octaedro committed:
Added bottom border to panel row title

octaedro committed:
Changed ActivityCard, now the children aren't required

octaedro committed:
Fixed button styles

octaedro committed:
Fixed title padding

octaedro committed:
Fixed order card in activity panel

octaedro committed:
Fixed orders number presentation when zero

octaedro committed:
Fixed layout.js conditional

octaedro committed:
ActivityPanel component refactored

octaedro committed:
Styles fixed

octaedro committed:
Added ActivityPanel tests

octaedro committed:
Removed orders from header

octaedro committed:
Fixed panel card styles

octaedro committed:
Fixed README.md

samueljseay commented on client/homescreen/activity-panel/orders/utils.js:
```diff
@@ -0,0 +1,45 @@
+
+	// Disable eslint rule requiring `latestNote` to be defined below because the next two statements
+	// depend on `getItemsTotalCount` to have been called.
+	// eslint-disable-next-line @wordpress/no-unused-vars-before-return
```
What do you think about refactoring this a little bit and moving it out of this area? It really should be a selector inside `@woocommerce/data` (in the items store). The only thing that will need to change to accomodate that is to pass in the `orderStatuses` to the selector, that way it's not dependent on the settings store directly to work.

Ideally we would also cover some of this with tests although that will probably require significant mocking right now, so optional imo, but quite important I think to start moving hidden bits of code like this to packages where they belong because it should be reusable.

octaedro replied:
I like the idea, if it's ok with you I will do this refactor in a follow-up PR.

octaedro replied:
Hi @samueljseay,
In order to address this fix, I created the commit 558cdd9, could you give it a look?

samueljseay commented on client/homescreen/activity-panel/panels.js:
```diff
@@ -0,0 +1,24 @@
+		},
+		// Add another panel row here
+	];
+	return panels;
```
Tiny thing but the variable declaration here is not needed. Just return the array directly and we can still add items to it easily.

octaedro replied:
This was changed in the commit 287e3cc

samueljseay commented on client/homescreen/activity-panel/style.scss:
```diff
@@ -0,0 +1,73 @@
+	}
+}
+
+.components-panel__body-title {
```
Any concerns with this not being scoped to the panel card? I could see this causing style pollution in other places. I realise that sometimes its necessary but I wanted to check.

octaedro replied:
Good eyes! I fixed it in the commit 1dbcfec

samueljseay commented on client/homescreen/activity-panel/test/index.js:
```diff
@@ -0,0 +1,58 @@
+	};
+} );
+
+// Mock the panels.
```
This comment should be "mock the orders"?

octaedro replied:
Fixed in the commit a5dd3fc

samueljseay commented on client/homescreen/activity-panel/test/index.js:
```diff
@@ -0,0 +1,58 @@
+import ActivityPanel from '../';
+
+// Mock the panels.
+jest.mock( '../panels', () => {
```
This mock reveals a code smell in the implementation of `<ActivityPanel>`. I can see why you're doing it, because if someone adds more rows then the test will fail, but it shows that you're using a private method inside `<ActivityPanel>` to render panels and that they should be a separate component.

Here's what I suggest:

1. Lets move away from the pattern of creating new objects like these panel objects which actually just represents what should be a component. Create a new `<Panel>` component.

A better panel component would be something like:

```
export const Panel = ({title, count, initialOpen, children}) => {

  // ... render the panel here
}
```

Some notes about why:

* every time we make a private method to render children in a component we remove a possible component we can reuse elsewhere. In the implementation of `<ActivityPanel>` as it is now, it cannot be reused and nor can any code used to generate the panels.

* Testing is easier, no mocking needed, test each very simple component separately.

* Note that increasing reuse will over time reduce bundle size because I imagine we're gonna reuse this UI pattern

2. With the new panel component in place you can now have `<ActivityPanel>` render it's children instead. Again making that component more flexible and removing the need for inlining component declarations in objects like we do in `getAllPanels`

3. Once `<ActivityPanel>` renders children we can write ActivityPanel like:

```
<ActivityPanel>
   {theData.map((whatever) => <Panel {...whatever} />)}
</ActivityPanel>
```

Doesn't look like much but in doing so we have decoupled the panel data from `ActivityPanel` (its just UI it doesn't need to know about the *data*.) And we made all the components fully reusable for elsewhere.

octaedro replied:
This structure was changed in the [PR 5474](https://github.com/woocommerce/woocommerce-admin/pull/5474)

samueljseay commented on client/homescreen/layout.js:
```diff
@@ -25,6 +25,7 @@ import '../dashboard/style.scss';
 import TaskListPlaceholder from '../task-list/placeholder';
 import InboxPanel from '../inbox-panel';
 import { WelcomeModal } from './welcome-modal';
+import ActivityPanel from './activity-panel';
```
Could we move towards always using named exports from now on? Its error prone to use defaults because its very easy to rename something different to what it was exported as and that makes it much harder to find usages in your IDE.

samueljseay commented on client/homescreen/activity-panel/index.js:
```diff
@@ -0,0 +1,59 @@
+	return <Panel> { renderPanels() } </Panel>;
+};
+
+export default withSelect( ( select ) => {
```
So i mentioned some code smells with this component in the tests but if you change those things this should also change.

If you want to pass state into `ActivityPanel` you will be able to pass it as props from its parent which is better because we separate the part thats just rendering from the part that should know about state.

samueljseay replied:
One other comment, I think we should consider that any HOC that gets state should use hooks from now on rather than using `compose` and `withSelect`. It makes sense to choose one approach and stick to it for consistency.

octaedro replied:
> One other comment, I think we should consider that any HOC that gets state should use hooks from now on rather than using compose and withSelect. It makes sense to choose one approach and stick to it for consistency.

This was modified in the commit a0c23d2

samueljseay reviewed (changes requested):
It's looking good, but I think we can improve reuse of the components built here. Have left some comments on how we can achieve that but reach out if you have any questions

samueljseay commented:
@octaedro One more thing thats been mentioned in discussions is that we should promote this new "accordion" component to `@woocommerce/components`. 

I know its a bit of a pain but I'm thinking it might make sense to split off a new PR for that component on its own within the components package and then this PR can follow up by consuming it.

Let me know if you have any issues with that approach. [Relevant discussion here](pbIJXs-ma-p2#comment-861)

octaedro committed:
Spelling error fixed

octaedro committed:
Moving orders panel

In progress

# Conflicts:
#	client/header/activity-panel/index.js

octaedro committed:
Moved Accordion component

This commit moves the "Accordion" component to "packages"

octaedro committed:
Removed const from panels.js

octaedro committed:
Fixed code comment

octaedro committed:
Added hook and changed exportation for ActivityPanel

This commit adds a hook and changes the exportation of the ActivityPanel

octaedro committed:
Added specificity to applied style

This commit adds specificity to the applied style

octaedro committed:
Added a fixed maximum of orders to show

octaedro commented:
Thank you for your review @samueljseay. Could you give this PR another look?

octaedro commented on client/header/activity-panel/index.js:
```diff
@@ -143,7 +137,7 @@ export class ActivityPanel extends Component {
 			! isPerformingSetupTask;
 
-		const showOrdersStockAndReviews =
+		const showstockAndReviews =
```
Nice catch! This was changed in the commit 2dfb29f

samueljseay commented on client/homescreen/activity-panel/index.js:
```diff
@@ -0,0 +1,46 @@
+						title,
+					} = panelData;
+					return (
+						<AccordionPanel
```
Yaaas, I don't know about you, but I love this!

samueljseay commented on client/homescreen/activity-panel/index.js:
```diff
@@ -0,0 +1,46 @@
+					} = panelData;
+					return (
+						<AccordionPanel
+							key={ index }
```
Not suggesting a change here but I just wanted to note that its *usually* a bad idea to use the index as the key. I think in this case its ok because we don't add/remove/reorder the list of panels, but its worth mentioning that if you were doing that you should have a stable unique ID for each item to use as the key. This video explains it well: https://egghead.io/lessons/react-use-the-key-prop-when-rendering-a-list-with-react

octaedro replied:
Good idea! This was changed in the commit 0d10549

octaedro committed:
Changed the copy displayed at Analytics

This commit changes the copy displayed at Analytics

octaedro committed:
Removed "index" from map()

octaedro commented:
Hi @samueljseay,
I rebased this PR and applied the change you suggested. Could you take a look at this PR again?

jeffstieler reviewed (dismissed):
Considering this is targeting `main` and isn't behind a feature flag, I think some visual/design items should be taken care of on this branch.

**Design:**
![68747470733a2f2f696d616765732e7a656e68756275736572636f6e74656e742e636f6d2f3564383466653865313230316465303030316236616236312f39653961663735622d356661382d343532332d396439652d316163303337346464326539](https://user-images.githubusercontent.com/63922/97597986-cf848480-19dc-11eb-9794-adf295780148.png)

**Implementd:**
![Screen Shot 2020-10-29 at 11 41 51 AM](https://user-images.githubusercontent.com/63922/97597987-cf848480-19dc-11eb-956c-d9aab7f05dc8.png)

Based on the design, the title for the card should be "Orders to fulfill" and the count should be to the left of the title. There's also a small visual issue with larger numbers (see 101 in my screenshot).

We should fix the "manage all orders" button before landing this.

jeffstieler commented:
I'm not sure how you'd like to prioritize this work @samueljseay, but we're going to need to rework the orders fetching logic as well. At a minimum to get rid of the "no orders to fulfill" as the default before even making a request for them:

![Capture 2020-10-29 at 12 50 28](https://user-images.githubusercontent.com/63922/97605921-9ac8fb00-19e5-11eb-955d-ca8996dafb59.gif)

The old Orders tab and panel made a few requests - one to power the "unread indicator" and one to retrieve the orders for display. I think that made sense when a user interaction was required to open the panel - while that was 2 requests, it potentially saved work by not retrieving orders before they were needed. The interaction is different now, so we should rework the data retrieval to suit.

octaedro committed:
Home Screen - Added empty activity panel for orders (#5467)

* Added redesign of empty card

* Added empty activity panel for orders

This commit adds an empty activity panel for orders

# Conflicts:
#	client/homescreen/activity-panel/panels.js

* Fixed OrdersPanel prop

# Conflicts:
#	client/homescreen/activity-panel/panels.js

* Removed hasNonActionableOrders from orders

* Added test

* Fixed Styles

Co-authored-by: Fernando Marichal <contacto@fernandomarichal.com>

octaedro committed:
Modified Accordion styles (#5506)

This commit modifies a few Accordion styles when toggling

Co-authored-by: Fernando Marichal <contacto@fernandomarichal.com>

octaedro committed:
Home Screen - Changed activity panel button styles (#5496)

* Fixed "Manage all orders" button

* Removed ActivityOutboundLink component

This commit removes the ActivityOutboundLink component

Co-authored-by: Fernando Marichal <contacto@fernandomarichal.com>

octaedro committed:
Removed ActivityOutboundLink component

octaedro committed:
Refactored panel outbound link

octaedro committed:
Fixed getUnreadOrders

This commit fixes the getUnreadOrders requesting

samueljseay reviewed (approved):
Pre approving this, I think you're ready to close this soon? amazing work @octaedro thank you!

🚀

octaedro committed:
Home Screen - Changed design and onClick behavior (#5516)

* Changed design and onClick behavior

This commit changes the design and onClick behavior for order cards

# Conflicts:
#	client/homescreen/activity-panel/orders/index.js
#	client/homescreen/activity-panel/orders/style.scss

# Conflicts:
#	client/homescreen/activity-panel/orders/index.js

* Update client/header/activity-panel/activity-card/index.js

Co-authored-by: Sam Seay <samueljseay@gmail.com>

* Added variable deconstruction

This commit adds a variable deconstruction

* Fixed activity card style

Co-authored-by: Fernando Marichal <contacto@fernandomarichal.com>
Co-authored-by: Sam Seay <samueljseay@gmail.com>

octaedro committed:
Refactored orders panel

octaedro committed:
Merge branch 'main' into add/5238

octaedro committed:
Fixed styles

octaedro committed:
Fixed flickering
